packages/nouns-contracts/contracts/libs/Noracle.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-3.0
+
+/// @title A library used to maintain Nouns Auction House price history
+
+/*********************************
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░██░░░████░░██░░░████░░░ *
+ * ░░██████░░░████████░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ *********************************/
+
+pragma solidity ^0.8.6;
+
+library Noracle {
+    uint32 public constant WARMUP_TIMESTAMP = 1;
+
+    error AlreadyInitialized();
+    error NotInitialized();
+    error AuctionCountOutOfBounds(uint32 auctionCount, uint32 cardinality);
+
+    struct Observation {
+        // The block.timestamp when the auction was settled
+        uint32 blockTimestamp;
+        // ID for the Noun (ERC721 token ID)
+        uint16 nounId;
+        // The winning bid amount, with 8 decimal places (reducing accuracy to save bits)
+        uint48 amount;
+        // The address of the auction winner
+        address winner;
+    }
+
+    struct NoracleState {
+        mapping(uint32 => Observation) observations;
+        uint32 index;
+        uint32 cardinality;
+        uint32 cardinalityNext;
+    }
+
+    function initialize(NoracleState storage self) internal {
+        if (self.cardinality > 0) revert AlreadyInitialized();
+
+        self.cardinality = 1;
+        self.cardinalityNext = 1;
+        warmUpObservation(self.observations[0]);
+    }
+
+    function write(
+        NoracleState storage self,
+        uint32 blockTimestamp,
+        uint16 nounId,
+        uint48 amount,
+        address winner
+    ) internal {
+        uint32 cardinalityNext = self.cardinalityNext;
+        if (cardinalityNext == 0) revert NotInitialized();
```

Comment:
maybe we can return instead of reverting?
the auction house can function without the oracle being initialized